doc/adr/0004-nats-headers.md

```diff
@@ -96,41 +85,22 @@ Use of headers is possible.
 ## Compatibility Across NATS Clients
 
 The following is a list of features to insure compatibility across NATS clients that support headers.
-Because the feature in Go client and nats-server leverage the Go implementation as described above,
-the API used will determine how header names are serialized.
 
-The Go implementation of `http.Header` is a `map[string][]string` If the `Set`, `Get`, `Values`, `Del`
-functions are used, the specified key arguments will be converted into their canonical format. This means
-that the serialized Header name will _not_ be what the application specified. Go's documentation
-motivates direct map access if the client doesn't wish to have the header modified. This has the implication
-that non-canonical Headers are invisible to `http.Header`'s API, as the API performs a canonical
-key conversion prior to any operation.
+In order to promote compatibility across clients, this section describes how NATS clients should behave:
 
-In order to promote compatibility across clients, this section describes how other language clients
-should behave:
-
-- `GET` returns a `string` of the first value found matching the specified key in a case-insensitive lookup.
-- `SET` removes all case-insensitive matching headers, with the specified value stored under the key 
-  specified without any manipulation.
-- `VALUES` returns a list of all values that case-insensitively match the specified key.
-- `DELETE` removes all headers that match case-insensitively the specified key.
-
-To prevent over-designing the API, it should be noted that if clients must have case-sensitive semantics,
-they should be able to iterate over the keys and values to enable client-side filtering of the values as
-necessary.
-
-#### Go Client
-
-Note that built-in Go client does not implement the above manipulations correctly. Users requiring
-case-preservation on a headers can easily implement these operations against the underlying map.
+- `GET` can either return a `string` of the first value found matching the specified key in a case-sensitive lookup, or it can return a list of `string`. This may vary depending on the client.
+- `VALUES` returns a list of all `string` values that case-sensitively match the specified key. This will not be present in clients where the `GET` functions like `VALUES`
+- `SET` removes all values stored under the case-sensitive key and replaces it with the new value. Some clients may refer to this a `PUT`
+- `ADD` If the case-sensitive key already has associated values, the new value is added to the list, otherwise the key is created with the new value.
+- `DELETE` removes all headers that match the specified case-sensitive key. Some clients may refer to this as `REMOVE`
+- Clients will provide case-insensitve versions of lookups such as `GET` and `VALUE`
```

Comment:
I think what could work is for `Get` to first match case-sensitive so that it becomes the fast path without relying on a normalization into canonical (http v1 style) or lowercase (http v2 style), that would help us with processing of JS headers where the case is already determined, but then if there is no match then do the case-insensitive check against the keys in the map?

Reply:
So I would ask why would we want to invent new computing behavior here when case paradigm is already baked into our languages and developers are used to it.

Seems to me that in support of case sensitive, is that fact that internally we specify an exact case for our internal header keys. In the corporate world, they tend to setup constants and organizational rules. http1 / 2 actually specify case sensitive keys with case-insensitive lookup, that's why I want to ensure clients provide the secondary ability to read case-insensitive

Reply:
You are right, for example things like `Sec-WebSocket-X` headers in HTTP actually have a correct way to be set so that is why in Go the `Set/Add` APIs aren't used for some things, otherwise it would become `Sec-Websocket-X` which can be incompatible in some proxies, for best compatibility instead it is set with the exact case via:

```go
req.Header["Sec-WebSocket-Key"]
```

The Go HTTP client does emit keys as they are set exactly when making a request as well to be flexible in that way without normalizing:

```go
	req, err := http.NewRequest("GET", "http://localhost:3000", nil)
	req.Header = http.Header{
		"foo": []string{"A"},
		"Foo": []string{"B"},
		"FOO": []string{"C"},
	}
	// RESULT:
	// User-Agent: Go-http-client/1.1
	// FOO: C
	// Foo: B
	// foo: A
```

The NATS Server currently does not normalize the keys either and treats them opaquely, so it is data-preserving already in that sense as well (and for JS headers that means that for using them correctly need to have the exact case (e.g. to be right it has to be exactly `Nats-Msg-Id`).  Maybe rather than defining the exact Get/Set/Del API for all clients what is important is that there is a way for users to get the keys in case-sensitive and case-insensitive ways as needed and document how to do it.

Reply:
> Maybe rather than defining the exact Get/Set/Del API for all clients what is important is that there is a way for users to get the keys in case-sensitive and case-insensitive ways as needed and document how to do it.
+1 I'm good with that